### PR TITLE
Ligand sdf round trip

### LINKED ISSUE #530: Make ligand schema round trip to SDF easily
While developing the Epik state expander I noticed that our internal Ligand model does not round trip to SDF easily. Consider this example case where I want to create a single SDF file with a list of ligands with their attributes such as `compound_name` added as SD tags.

```python
from asapdiscovery.data.openeye import save_openeye_sdfs, load_openeye_sdfs, oemol_to_sdf_string
from asapdiscovery.data.schema_v2.ligand import Ligand

ligands = [Ligand.from_smiles("CC", compound_name="ethane", tags={"parent":"me"}, experimental_data={"compound_id":"1"}), Ligand.from_smiles("CO", compound_name="methanol", ids={"moonshot_compound_id": "randomID"})]

# flush the data
for ligand in ligands:
    # sync the state internal and sdf data
    ligand.data = ligand.flush_attrs_to_SD_data()

oe_mols = [ligand.to_oemol() for ligand in ligands]

# write to sdf
save_openeye_sdfs(oe_mols, "test.sdf")

# load back
loaded_mols = load_openeye_sdfs("test.sdf")
# we need each mol to be an sdf to read with attrs
new_ligands = []
for ligand in loaded_mols:
    with open("ligand.sdf", "w") as out:
        out.write(oemol_to_sdf_string(ligand))
    new_ligands.append(Ligand.from_sdf("ligand.sdf", read_SD_attrs=True))

for ligand in new_ligands:
    print(ligand)
```

What I suggest is that we make the model round trip to and from openeye by default so all tags are included and parsed everytime we convert to and from oemols. I would also only store the graph and coords data in the internal data field and have the sd tags added automatically everytime we convert to an oemol to keep the models in sync as its easy to forget to flush the data before writing to file.

cc @hmacdope

### PR DESCRIPTION
**Pull request by jthorton**
## Description
This PR fixes #530 by allowing our internal Ligand model to round trip to SDF files using SD tags to store extra information. We now store pydantic models as JSON strings in SD tags which even allows for nested models and is general so that any new models added to ligand will automatically survive the round trip. This will allow us to SDFs through the workflow without losing any important information. 

One notable change is that the internal copy of the sdf string we keep in the data field now has no SD tags, these are now added when ever we convert to openeye mols to keep the internal state of the molecule consistent. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] TODO 1

## Questions
- [ ] Question1

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by jthorton** [comment]
@hmacdope I think this is ready to go and should allow me to finally finish the expanders I hope!

File: asapdiscovery-data/asapdiscovery/data/openeye.py [hunk 1]
```diff
@@ -148,6 +148,44 @@ def load_openeye_pdb(
...
+    Load an OpenEye SMILES file containing a single molecule and return it as an
+    OpenEye OEGraphMol object.
+    Parameters
+    ----------
+    smi_fn : Union[str, Path]
+        Path to the SMILES file to load.
+    Returns
+    -------
+    oechem.OEGraphMol
+        An OpenEye OEGraphMol object containing the molecule data from the SDF file.
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/openeye.py:162]:
```
        An OpenEye OEGraphMol object containing the molecule data from the SDF file.
```

File: asapdiscovery-data/asapdiscovery/data/openeye.py [hunk 2]
```diff
@@ -148,6 +148,44 @@ def load_openeye_pdb(
         oechem.OEThrow.Fatal(f"Unable to open {pdb_fn}")
 
 
+def load_openeye_smi(smi_fn: Union[str, Path]) -> oechem.OEGraphMol:
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/openeye.py:151]: Single smiles only? probably better to have it return all the ones in a file, I know there is a bit of a historical artifact tho this, but still would probably prefer loading all the mols.

File: asapdiscovery-data/asapdiscovery/data/openeye.py [hunk 3]
```diff
@@ -148,6 +148,44 @@ def load_openeye_pdb(
...
+        An OpenEye OEGraphMol object containing the molecule data from the SDF file.
+    Raises
+    ------
+    FileNotFoundError
+        If the specified file does not exist.
+    oechem.OEError
+        If the CIF file cannot be opened.
+    Notes
+    -----
+    This function assumes that the SDF file contains a single molecule. If the
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/openeye.py:171]: See above.

File: asapdiscovery-data/asapdiscovery/data/openeye.py [hunk 4]
```diff
@@ -728,7 +804,7 @@ def _set_SD_data_repr(mol: oechem.OEMol, data: dict[str, Any]) -> oechem.OEMol:
         OpenEye OEMol with SD data set
     """
     # NOTE: use repr to ensure re-reading the SD data will give the same value
-    mol = set_SD_data(mol, {k: repr(v) for k, v in data.items()})
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/openeye.py:731]: Change comment above if changing this, might also want to wrap in try/except for string cast, unless guaranteed that all incoming objects will be castable.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/experimental.py [hunk 5]
```diff
@@ -47,13 +47,13 @@ class ExperimentalCompoundData(Model):
...
-        """
-        Convert to a dictionary of SD tags
-        """
-        data = self.dict()
-        exp_data = data.pop("experimental_data")
-        # cannot use a nested dict in SD tags, so flatten to two tags
-        data = {str(k): str(v) for k, v in data.items() if v is not None}
-        exp_data = {str(k): float(v) for k, v in exp_data.items() if v is not None}
-        return data, exp_data
+    # def to_SD_tags(self) -> tuple[dict[str, str], dict[str, float]]:
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/experimental.py:50]: remove commented out bits.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/identifiers.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        None, description="Unique ID from Postera Manifold API"
+    )
+    manifold_vc_id: Optional[str] = Field(
+        None, description="Unique VC ID (virtual compound ID) from Postera Manifold"
+    )
+    compchem_id: Optional[UUID4] = Field(
+        None, description="Unique ID for P5 compchem reference, unused for now"
+    )
+
+    # def to_SD_tags(self) -> dict[str, str]:
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/identifiers.py:38]: Same here remove commented out bits.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/identifiers.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
+from typing import Optional
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/identifiers.py:1]: If this is going to be a seperate file, move `TargetIdentifiers` here also.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py [hunk 8]
```diff
@@ -112,12 +73,12 @@ class Ligand(DataModelAbstractBase):
...
 
     data: str = Field(
-        "",
+        ...,
         description="SDF file stored as a string to hold internal data state",
         repr=False,
     )
-    data_format: DataStorageType = Field(
-        DataStorageType.sdf,
+    data_format: Literal["sdf"] = Field(
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:80]: Can we keep the enum? It should cast easily to a string anyway, but we can use `Literal`

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py [hunk 9]
```diff
@@ -311,60 +298,87 @@ def clear_SD_data(self) -> None:
...
-            )
-        # get reserved attribute names
-        reser_attr_names = [attr.name for attr in self.__fields__.values()]
-        # push all non reserved attribute names to tags
-        data["tags"].update(
-            {k: v for k, v in data.items() if k not in reser_attr_names}
-        )
-        # reinitialise object
-        self.__init__(**data)
+    # def _clear_internal_SD_data(self) -> None:
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:301]: Remove commented

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking really good! Just a few things to query. I think this is a much better solution than the one I had.

File: asapdiscovery-data/asapdiscovery/data/openeye.py (hunk 4 above)
Comment (jthorton) [reply asapdiscovery-data/asapdiscovery/data/openeye.py:731]: Good catch I think everything from the ligand side should be a string as we are enforcing it via typing and using json models so we should be safe to change the type hint to expect dicts of str only.

File: asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py (hunk 8 above)
Comment (jthorton) [reply asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py:80]: Found a way to do it just had to run it through JSON to get it to work!

**Comment by jthorton** [comment]
continued in #534

**Pull request closed without merging**